src/meta/meta_apply.c

```diff
@@ -10,55 +10,71 @@
 
 /*
  * __meta_btree_apply --
- *     Apply a function to all files listed in the metadata, apart from the metadata file.
+ *     Conditionally apply a function to the given file URI.
  */
 static inline int
-__meta_btree_apply(WT_SESSION_IMPL *session, WT_CURSOR *cursor,
+__meta_btree_apply(WT_SESSION_IMPL *session, const char *uri,
   int (*file_func)(WT_SESSION_IMPL *, const char *[]),
   int (*name_func)(WT_SESSION_IMPL *, const char *, bool *), const char *cfg[])
 {
     WT_DECL_RET;
     int t_ret;
-    const char *uri;
     bool skip;
 
+    skip = false;
+    if (name_func != NULL)
+        ret = name_func(session, uri, &skip);
+
+    if (file_func == NULL || skip || !WT_PREFIX_MATCH(uri, "file:"))
+        return (ret);
+
     /*
-     * Accumulate errors but continue through to the end of the metadata.
+     * We need to pull the handle into the session handle cache and make sure it's referenced to
+     * stop other internal code dropping the handle (e.g in LSM when cleaning up obsolete chunks).
+     * Holding the schema lock isn't enough.
+     *
+     * Handles that are busy are skipped without the whole operation failing. This deals among other
+     * cases with checkpoint encountering handles that are locked (e.g., for bulk loads or verify
+     * operations).
      */
-    while ((t_ret = cursor->next(cursor)) == 0) {
-        if ((t_ret = cursor->get_key(cursor, &uri)) != 0 || strcmp(uri, WT_METAFILE_URI) == 0) {
-            WT_TRET(t_ret);
-            continue;
-        }
+    if ((t_ret = __wt_session_get_dhandle(session, uri, NULL, NULL, 0)) != 0) {
```

Comment:
Is it really fine to just proceed even if this fails? In this prev code, we were doing `continue` to stop ourselves.

Reply:
Good point, fixed.